### PR TITLE
[SPARK-23900][SQL] format_number support user specifed format as argument

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

`format_number` support user specifed format as argument. For example:
```sql
spark-sql> SELECT format_number(12332.123456, '##################.###');
12332.123
```

## How was this patch tested?

unit test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -2022,6 +2022,8 @@ case class Encode(value: Expression, charset: Expression)
     Examples:
       > SELECT _FUNC_(12332.123456, 4);
        12,332.1235
+      > SELECT _FUNC_(12332.123456, '##################.###');
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: We need to update `ExpressionDescription` too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -2050,33 +2058,49 @@ case class FormatNumber(x: Expression, d: Expression)
...
+            numberFormat.applyLocalizedPattern(pattern.toString)
+        }
+      case StringType =>
+        val dValue = dObject.asInstanceOf[UTF8String].toString
+        lastDStringValue match {
+          case Some(last) if last == dValue =>
+          case _ =>
+            pattern.delete(0, pattern.length)
+            lastDStringValue = Some(dValue)
+            if (dValue.toString.isEmpty) {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: `dValue` is already a string.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -2108,35 +2132,52 @@ case class FormatNumber(x: Expression, d: Expression)
       // SPARK-13515: US Locale configures the DecimalFormat object to use a dot ('.')
       // as a decimal separator.
       val usLocale = "US"
-      val i = ctx.freshName("i")
       val dFormat = ctx.freshName("dFormat")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: `dFormat` is not used in any place. We can remove it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 4]
```diff
@@ -2108,35 +2132,52 @@ case class FormatNumber(x: Expression, d: Expression)
...
-      val lastDValue =
-        ctx.addMutableState(CodeGenerator.JAVA_INT, "lastDValue", v => s"$v = -100;")
       val pattern = ctx.addMutableState(sb, "pattern", v => s"$v = new $sb();")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: `pattern` is only needed for `IntegerType` case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 5]
```diff
@@ -2108,35 +2132,52 @@ case class FormatNumber(x: Expression, d: Expression)
...
-          ${ev.value} = null;
-          ${ev.isNull} = true;
-        }
-       """
+           """
+        case StringType =>
+          val lastDStringValue =
+            ctx.addMutableState("String", "lastDValue", v => s"""$v = "$defaultFormat";""")
+          s"""
+            if (!$d.toString().equals($lastDStringValue)) {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: We should only call `$d.toString()` once.

**Comment by gatorsmile** [comment]
cc @ueshin

**Comment by wangyum** [comment]
retest please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 6]
```diff
@@ -2108,35 +2133,53 @@ case class FormatNumber(x: Expression, d: Expression)
...
-        } else {
-          ${ev.value} = null;
-          ${ev.isNull} = true;
-        }
-       """
+           """
+        case StringType =>
+          val lastDStringValue =
+            ctx.addMutableState("String", "lastDValue", v => s"""$v = "$defaultFormat";""")
+          val dValue = ctx.addMutableState("String", "dValue")
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Do we need to make this mutable state?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 7]
```diff
@@ -2108,35 +2133,53 @@ case class FormatNumber(x: Expression, d: Expression)
...
+          s"""
+            $dValue = $d.toString();
+            if (!$dValue.equals($lastDStringValue)) {
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: What if the first `dValue` is the same as the default format? Can you add the test case?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 8]
```diff
@@ -2108,35 +2133,57 @@ case class FormatNumber(x: Expression, d: Expression)
...
-          ${ev.value} = UTF8String.fromString($numberFormat.format(${typeHelper(num)}));
-        } else {
-          ${ev.value} = null;
-          ${ev.isNull} = true;
-        }
-       """
+           """
+        case StringType =>
+          val lastDValue = ctx.addMutableState("String", "lastDValue", v => s"""$v = null;""")
+          val dValue = ctx.addMutableState("String", "dValue")
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Do we need to make this mutable state?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 9]
```diff
@@ -2108,35 +2133,57 @@ case class FormatNumber(x: Expression, d: Expression)
...
+              $lastDValue = $dValue;
+              if ($dValue.isEmpty()) {
+                $numberFormat.applyLocalizedPattern("$defaultFormat");
+              } else {
+                $numberFormat.applyLocalizedPattern($dValue);
+              }
+            }
+            ${ev.value} = UTF8String.fromString($numberFormat.format(${typeHelper(num)}));
+           """
+        case NullType =>
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: We don't need this case?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 10]
```diff
@@ -706,6 +706,30 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+        Literal("###,###,###,###,###.####")), "15,159,339,180,002,773.2778")
+    checkEvaluation(FormatNumber(Literal.create(null, IntegerType), Literal("##.###")), null)
+    assert(FormatNumber(Literal.create(null, NullType), Literal("##.###")).resolved === false)
+
+    checkEvaluation(FormatNumber(Literal(12332.123456), Literal("#,###,###,###,###,###,##0")),
+      "12,332")
+    checkEvaluation(FormatNumber(
+      Literal.create(null, IntegerType), Literal.create(null, NullType)), null)
+    checkEvaluation(FormatNumber(
+      Literal.create(null, NullType), Literal.create(null, NullType)), null)
```
Comment (ueshin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: What's for these two cases?

**Comment by HyukjinKwon** [comment]
ping @wangyum

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by ueshin** [comment]
Basically LGTM, but I'm wondering what if the `expr2` is not like a format string?

**Comment by wangyum** [comment]
> Basically LGTM, but I'm wondering what if the expr2 is not like a format string?

The same as Hive:
```sql
spark-sql> SELECT format_number(12332.123456, 'abc');
abc12332
```
```sql
hive> SELECT format_number(12332.123456, 'abc');
OK
abc12332
Time taken: 0.218 seconds, Fetched: 1 row(s)
```

**Comment by ueshin** [comment]
Thanks, sounds good.
Let me retrigger the build just for checking again.

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by wangyum** [comment]
Jenkins, retest this please.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**